shadowsocks-csharp/Controller/System/ProtocolHandler.cs

```diff
@@ -0,0 +1,104 @@
+﻿using Microsoft.Win32;
+using NLog;
+using Shadowsocks.Util;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Shadowsocks.Controller
+{
+    static class ProtocolHandler
+    {
+        private static Logger logger = LogManager.GetCurrentClassLogger();
+
+        // Don't use Application.ExecutablePath
+        // see https://stackoverflow.com/questions/12945805/odd-c-sharp-path-issue
+        private static readonly string ExecutablePath = Assembly.GetEntryAssembly().Location;
+
+        // TODO: Elevate when necessary
+        public static bool Set(bool enabled)
+        {
+            RegistryKey ssURLAssociation = null;
+            try
+            {
+                ssURLAssociation = Registry.CurrentUser.CreateSubKey(@"SOFTWARE\Classes\ss", RegistryKeyPermissionCheck.ReadWriteSubTree);
```

Comment:
Perhaps using full name "Shadowsocks" is better

Reply:
> Perhaps using full name "Shadowsocks" is better

By using "Shadowsocks" you are registering Shadowsocks:// instead of ss://.

Reply:
Thanks @database64128 , I made a mistake...

Note:
Outline or other software may also register the protocol
![image](https://user-images.githubusercontent.com/3608762/79681609-2f3fed00-824e-11ea-99f7-34c235e0f8ee.png)

Reply:
> Outline or other software may also register the protocol

Just tested with Outline installed. The current behavior of our implementation is overwriting existing association, which is not perfect, but good enough. As far as I know, there is no way to associate a URI scheme with more than one Win32 application. The [API for that](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/handle-uri-activation) is only available for UWP apps.

Reply:
Also note that Outline overwrites the URI association on boot. Apparently they are using the same implementation, with a more aggressive approach and no way to turn off.

Reply:
Our implementation checks if ss:// is associated with itself on boot to determine the status of association, and shows it with `MenuItem.Checked`. It's up to the user to associate the URI scheme or not.

Reply:
> with a more aggressive approach and no way to turn off

We need open an issue for it.